steel/src/state.rs

```diff
@@ -0,0 +1,207 @@
+// Copyright 2024 RISC Zero, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{convert::Infallible, fmt::Debug, rc::Rc};
+
+use crate::mpt::{MerkleTrie, EMPTY_ROOT_HASH};
+use alloy_primitives::{keccak256, Address, Bytes, TxNumber, B256, U256};
+use alloy_rlp_derive::{RlpDecodable, RlpEncodable};
+use revm::{
+    primitives::{AccountInfo, Bytecode, HashMap, KECCAK_EMPTY},
+    Database,
+};
+
+/// A simple MPT-based read-only EVM database implementation.
+///
+/// It is backed by a single [MerkleTrie] for the accounts and one [MerkleTrie] each for the
+/// accounts' stores. It panics when querying data not contained in the tries. This design allows
+/// storage keys to be queried by storage trie root hash, but not by account, which is required to
+/// implement [revm::DatabaseRef]. Thus, in order to use [StateDb] in revm, [WrapStateDb] must be
+/// used, which caches the appropriate storage trie root for each basic account query, thus
+/// requiring mutability.
+pub struct StateDb {
+    /// State MPT.
+    state_trie: MerkleTrie,
+    /// Storage MPTs to their root hash.
+    /// [Rc] is used fore MPT deduplication.
+    storage_tries: HashMap<B256, Rc<MerkleTrie>>,
+    /// Contracts by their hash.
+    contracts: HashMap<B256, Bytes>,
+    /// Block hashes by their number.
+    block_hashes: HashMap<u64, B256>,
```

Comment:
We might want to use `BTreeMap` here. I've seen it perform much better in the zkVM.

Reply:
I did some quick Benchmarks and at least for this very particular usecase where the `block_hashes` never contains more than 256 elements. The performance was more or less identical. The initialisation of the HashMap was a tiny bit faster while the getting was a tiny bit faster for the BTreeMap.
I guess in this case I'll leave it at the `HashMap` because of the better theoretical performance.

Reply:
Ya, I don't think it'll matter either way then here 👍